File: lib/python/opentoken-cli/src/test/opentoken_cli/main_test.py
"""
Copyright (c) Truveta. All rights reserved.

Integration tests for the main module.
Tests the end-to-end workflows for token generation and decryption using new subcommand interface.
"""

import os
import pytest
import tempfile
from pathlib import Path

from opentoken_cli.commands.open_token_command import OpenTokenCommand


class TestOpenTokenCommand:
    """Integration tests for OpenToken CLI with new subcommand interface."""

    HASHING_SECRET = "TestHashingSecret"
    ENCRYPTION_KEY = "TestEncryptionKeyValue1234567890"  # Must be exactly 32 chars

    @pytest.fixture
    def temp_dir(self, tmp_path):
        """Create temporary directory with test input CSV."""
        input_csv = tmp_path / "input.csv"
        csv_content = (
            "RecordId,FirstName,LastName,PostalCode,Sex,BirthDate,SocialSecurityNumber\n"
            "test-001,John,Doe,98004,Male,2000-01-01,123-45-6789\n"
            "test-002,Jane,Smith,12345,Female,1990-05-15,234-56-7890\n"
        )
        input_csv.write_text(csv_content)
        return tmp_path

    def test_package_command_csv_to_csv(self, temp_dir):
        """Test package command (tokenize + encrypt) with CSV input/output."""
        input_csv = temp_dir / "input.csv"
        output_csv = temp_dir / "output.csv"

        args = [
            "package",
            "-i", str(input_csv),
            "-t", "csv",
            "-o", str(output_csv),
            "--hashingsecret", self.HASHING_SECRET,
            "--encryptionkey", self.ENCRYPTION_KEY
        ]

        exit_code = OpenTokenCommand.execute(args)

        assert exit_code == 0, "Command should execute successfully"
        assert output_csv.exists(), "Output CSV should be created"
        assert output_csv.stat().st_size > 0, "Output CSV should not be empty"

        # Check metadata file
        metadata_path = temp_dir / "output.metadata.json"
        assert metadata_path.exists(), "Metadata file should be created"

    def test_package_command_csv_to_parquet(self, temp_dir):
        """Test package command with CSV input and Parquet output."""
        input_csv = temp_dir / "input.csv"
        output_parquet = temp_dir / "output.parquet"

        args = [
            "package",
            "-i", str(input_csv),
            "-t", "csv",
            "-o", str(output_parquet),
            "-ot", "parquet",
            "--hashingsecret", self.HASHING_SECRET,
            "--encryptionkey", self.ENCRYPTION_KEY
        ]

        exit_code = OpenTokenCommand.execute(args)

        assert exit_code == 0, "Command should execute successfully"
        assert output_parquet.exists(), "Output Parquet should be created"
        assert output_parquet.stat().st_size > 0, "Output Parquet should not be empty"

    def test_tokenize_command(self, temp_dir):
        """Test tokenize command (hash-only, no encryption)."""
        input_csv = temp_dir / "input.csv"
        output_csv = temp_dir / "output.csv"

        args = [
            "tokenize",
            "-i", str(input_csv),
            "-t", "csv",
            "-o", str(output_csv),
            "--hashingsecret", self.HASHING_SECRET
        ]

        exit_code = OpenTokenCommand.execute(args)

        assert exit_code == 0, "Command should execute successfully"
        assert output_csv.exists(), "Output CSV should be created"
        assert output_csv.stat().st_size > 0, "Output CSV should not be empty"

    def test_decrypt_command(self, temp_dir):
        """Test decrypt command."""
        input_csv = temp_dir / "input.csv"
        output_csv = temp_dir / "output.csv"
        decrypted_csv = temp_dir / "decrypted.csv"

        # First, generate encrypted tokens
        encrypt_args = [
            "package",
            "-i", str(input_csv),
            "-t", "csv",
            "-o", str(output_csv),
            "--hashingsecret", self.HASHING_SECRET,
            "--encryptionkey", self.ENCRYPTION_KEY
        ]
        OpenTokenCommand.execute(encrypt_args)

        # Now decrypt them
        decrypt_args = [
            "decrypt",
            "-i", str(output_csv),
            "-t", "csv",
            "-o", str(decrypted_csv),
            "--encryptionkey", self.ENCRYPTION_KEY
        ]

        exit_code = OpenTokenCommand.execute(decrypt_args)

        assert exit_code == 0, "Command should execute successfully"
        assert decrypted_csv.exists(), "Decrypted CSV should be created"
        assert decrypted_csv.stat().st_size > 0, "Decrypted CSV should not be empty"

    def test_output_type_defaults_to_input_type(self, temp_dir):
        """Test that output type defaults to input type when not specified."""
        input_csv = temp_dir / "input.csv"
        output_csv = temp_dir / "output.csv"

        args = [
            "package",
            "-i", str(input_csv),
            "-t", "csv",
            "-o", str(output_csv),
            "--hashingsecret", self.HASHING_SECRET,
            "--encryptionkey", self.ENCRYPTION_KEY
        ]

        exit_code = OpenTokenCommand.execute(args)

        assert exit_code == 0, "Command should execute successfully"
        assert output_csv.exists(), "Output CSV should be created"

        # Verify CSV output was created (same as input type)
        content = output_csv.read_text()
        assert "RecordId" in content, "Output should contain CSV headers"

    def test_parquet_input_to_parquet_output(self, temp_dir):
        """Test Parquet input to Parquet output."""
        input_csv = temp_dir / "input.csv"
        temp_parquet = temp_dir / "temp.parquet"
        output_parquet = temp_dir / "output2.parquet"

        # First create a parquet file from CSV
        create_args = [
            "package",
            "-i", str(input_csv),
            "-t", "csv",
            "-o", str(temp_parquet),
            "-ot", "parquet",
            "--hashingsecret", self.HASHING_SECRET,
            "--encryptionkey", self.ENCRYPTION_KEY
        ]
        OpenTokenCommand.execute(create_args)

        # Now use parquet as input
        args = [
            "decrypt",
            "-i", str(temp_parquet),
            "-t", "parquet",
            "-o", str(output_parquet),
            "--encryptionkey", self.ENCRYPTION_KEY
        ]

        exit_code = OpenTokenCommand.execute(args)
        assert exit_code == 0, "Command should execute successfully"

    def test_decrypt_csv_to_parquet(self, temp_dir):
        """Test decrypting CSV to Parquet format."""
        input_csv = temp_dir / "input.csv"
        output_csv = temp_dir / "output.csv"
        decrypted_parquet = temp_dir / "decrypted.parquet"

        # First generate encrypted tokens
        encrypt_args = [
            "package",
            "-i", str(input_csv),
            "-t", "csv",
            "-o", str(output_csv),
            "--hashingsecret", self.HASHING_SECRET,
            "--encryptionkey", self.ENCRYPTION_KEY
        ]
        OpenTokenCommand.execute(encrypt_args)

        # Decrypt CSV to Parquet
        decrypt_args = [
            "decrypt",
            "-i", str(output_csv),
            "-t", "csv",
            "-o", str(decrypted_parquet),
            "-ot", "parquet",
            "--encryptionkey", self.ENCRYPTION_KEY
        ]

        exit_code = OpenTokenCommand.execute(decrypt_args)
        assert exit_code == 0, "Command should execute successfully"
        assert decrypted_parquet.exists(), "Decrypted Parquet should be created"

File: lib/java/opentoken-cli/src/test/java/com/truveta/opentoken/cli/MainTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.cli;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.io.PrintStream;
import java.nio.file.Files;
import java.nio.file.Path;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import com.truveta.opentoken.cli.commands.OpenTokenCommand;

/**
 * Integration tests for {@link Main} class.
 * Tests the end-to-end workflows for token generation and decryption using new subcommand interface.
 */
class MainTest {

    @TempDir
    Path tempDir;

    private Path inputCsv;
    private Path outputCsv;
    private Path outputParquet;
    private PrintStream originalErr;

    private static final String HASHING_SECRET = "TestHashingSecret";
    private static final String ENCRYPTION_KEY = "TestEncryptionKeyValue1234567890"; // Must be exactly 32 chars

    @BeforeEach
    void setUp() throws IOException {
        // Create test input CSV
        inputCsv = tempDir.resolve("input.csv");
        outputCsv = tempDir.resolve("output.csv");
        outputParquet = tempDir.resolve("output.parquet");

        String csvContent = "RecordId,FirstName,LastName,PostalCode,Sex,BirthDate,SocialSecurityNumber\n"
                + "test-001,John,Doe,98004,Male,2000-01-01,123-45-6789\n"
                + "test-002,Jane,Smith,12345,Female,1990-05-15,234-56-7890\n";
        Files.writeString(inputCsv, csvContent);

        // Capture stderr for error logging verification
        originalErr = System.err;
    }

    @AfterEach
    void tearDown() {
        System.setErr(originalErr);
    }

    @Test
    void testPackageCommandCsvToCsv() throws IOException {
        String[] args = {
                "package",
                "-i", inputCsv.toString(),
                "-t", "csv",
                "-o", outputCsv.toString(),
                "--hashingsecret", HASHING_SECRET,
                "--encryptionkey", ENCRYPTION_KEY
        };

        assertDoesNotThrow(() -> OpenTokenCommand.execute(args));

        assertTrue(Files.exists(outputCsv), "Output CSV should be created");
        assertTrue(Files.size(outputCsv) > 0, "Output CSV should not be empty");

        // Check metadata file
        Path metadataPath = tempDir.resolve("output.metadata.json");
        assertTrue(Files.exists(metadataPath), "Metadata file should be created");
    }

    @Test
    void testPackageCommandCsvToParquet() throws IOException {
        String[] args = {
                "package",
                "-i", inputCsv.toString(),
                "-t", "csv",
                "-o", outputParquet.toString(),
                "-ot", "parquet",
                "--hashingsecret", HASHING_SECRET,
                "--encryptionkey", ENCRYPTION_KEY
        };

        assertDoesNotThrow(() -> OpenTokenCommand.execute(args));

        assertTrue(Files.exists(outputParquet), "Output Parquet should be created");
        assertTrue(Files.size(outputParquet) > 0, "Output Parquet should not be empty");
    }

    @Test
    void testTokenizeCommand() throws IOException {
        String[] args = {
                "tokenize",
                "-i", inputCsv.toString(),
                "-t", "csv",
                "-o", outputCsv.toString(),
                "--hashingsecret", HASHING_SECRET
        };

        assertDoesNotThrow(() -> OpenTokenCommand.execute(args));

        assertTrue(Files.exists(outputCsv), "Output CSV should be created");
        assertTrue(Files.size(outputCsv) > 0, "Output CSV should not be empty");
    }

    @Test
    void testDecryptCommand() throws IOException {
        // First, generate encrypted tokens
        String[] encryptArgs = {
                "package",
                "-i", inputCsv.toString(),
                "-t", "csv",
                "-o", outputCsv.toString(),
                "--hashingsecret", HASHING_SECRET,
                "--encryptionkey", ENCRYPTION_KEY
        };
        OpenTokenCommand.execute(encryptArgs);

        // Now decrypt them
        Path decryptedCsv = tempDir.resolve("decrypted.csv");
        String[] decryptArgs = {
                "decrypt",
                "-i", outputCsv.toString(),
                "-t", "csv",
                "-o", decryptedCsv.toString(),
                "--encryptionkey", ENCRYPTION_KEY
        };

        assertDoesNotThrow(() -> OpenTokenCommand.execute(decryptArgs));

        assertTrue(Files.exists(decryptedCsv), "Decrypted CSV should be created");
        assertTrue(Files.size(decryptedCsv) > 0, "Decrypted CSV should not be empty");
    }

    @Test
    void testOutputTypeDefaultsToInputType() throws IOException {
        String[] args = {
                "package",
                "-i", inputCsv.toString(),
                "-t", "csv",
                "-o", outputCsv.toString(),
                "--hashingsecret", HASHING_SECRET,
                "--encryptionkey", ENCRYPTION_KEY
        };

        assertDoesNotThrow(() -> OpenTokenCommand.execute(args));

        // Verify CSV output was created (same as input type)
        assertTrue(Files.exists(outputCsv));
        String content = Files.readString(outputCsv);
        assertTrue(content.contains("RecordId"));
    }

    @Test
    void testParquetInputToParquetOutput() throws IOException {
        // First create a parquet file from CSV
        Path tempParquet = tempDir.resolve("temp.parquet");
        String[] createArgs = {
                "package",
                "-i", inputCsv.toString(),
                "-t", "csv",
                "-o", tempParquet.toString(),
                "-ot", "parquet",
                "--hashingsecret", HASHING_SECRET,
                "--encryptionkey", ENCRYPTION_KEY
        };
        OpenTokenCommand.execute(createArgs);

        // Now use parquet as input
        Path outputParquet2 = tempDir.resolve("output2.parquet");
        String[] args = {
                "decrypt",
                "-i", tempParquet.toString(),
                "-t", "parquet",
                "-o", outputParquet2.toString(),
                "--encryptionkey", ENCRYPTION_KEY
        };

        assertDoesNotThrow(() -> OpenTokenCommand.execute(args));
    }

    @Test
    void testDecryptCsvToParquet() throws IOException {
        // First generate encrypted tokens
        String[] encryptArgs = {
                "package",
                "-i", inputCsv.toString(),
                "-t", "csv",
                "-o", outputCsv.toString(),
                "--hashingsecret", HASHING_SECRET,
                "--encryptionkey", ENCRYPTION_KEY
        };
        OpenTokenCommand.execute(encryptArgs);

        // Decrypt CSV to Parquet
        Path decryptedParquet = tempDir.resolve("decrypted.parquet");
        String[] decryptArgs = {
                "decrypt",
                "-i", outputCsv.toString(),
                "-t", "csv",
                "-o", decryptedParquet.toString(),
                "-ot", "parquet",
                "--encryptionkey", ENCRYPTION_KEY
        };

        assertDoesNotThrow(() -> OpenTokenCommand.execute(decryptArgs));
        assertTrue(Files.exists(decryptedParquet));
    }
}
